lib/fluent/plugin/kubernetes_metadata_common.rb

```diff
@@ -82,14 +82,29 @@ def parse_pod_metadata(pod_object)
         log.warn("parsing container meta information failed for: #{pod_object[:metadata][:namespace]}/#{pod_object[:metadata][:name]}: #{e}")
       end
 
+      ownerrefs_meta = {}
+      begin
+        key = 0
+        pod_object[:metadata][:ownerReferences]. each do |owner_reference|
+          ownerrefs_meta[key.to_s] = {
```

Comment:
Thinking about this more, why does this need to be a map where the key is a number?  Why can it not be implemented as an array? I don't see the need for map given its not a "named" thing

Reply:
I originally started it as an array, but I didn't like the data representation on downstream systems. When it got ingested into our OpenSearch store, it ended up being decoded as a JSON blob rather than a field. e.g. I would end up with something like:

```
kubernetes.pod_name = foo-123151-adf12411
kubernetes.ownerrefs = {"kind": "ReplicaSet", "name": "foo-123151"}
```

I found this much harder to work with. Many of our users are used to maps, because every other field that comes through fluentd is a map. Additionally, the above is a lot harder to search on or parse than the other mapped fields that the metadata plugin produces. Finally, in this method, it gives me a consistent way to reference the primary owner via search fields e.g. `kubernetes.ownerrefs.0.kind` or some such.

This may be somewhat specific to our use case. I'm not sure how all of the other downstream log stores would handle the above, I only know about the ElasticSearch / Kibana family of tools. However, I would guess that without a consistent key field a lot of the other tools would struggle here as well. In general I would say that since all of the other fields that we produce are maps, it is better to go this way for consistency.

Reply:
@aauren - I don't understand what you mean by "decode as JSON blob", can you illustrate with some more complete examples?

If I understand your map looks like:
```
{
  "kubernetes": {
    "controller": {
      "0": { "kind":"foo", ...}
      "1": { "kind":"bar", ...}
   }
}
```
And the path expression to extract the kind field looks like:
```
kubernetes.controller.0.kind
```

As an array this looks like:
```
{
  "kubernetes": {
    "controllers": [
      { "kind":"foo", ...}
      { "kind":"bar", ...}
   ]
}
```
And the path expression to extract the kind field looks like:
```
kubernetes.controller[0].kind
```

I don't see how the map version is easier to use or understand. Unless the integer keys mean something, I don't think they should be required. If controllers have a unique string name or ID then that would be a good map key, but if you want a collection of objects with no meaningful keys then I think an array is the way to go.

I may be missing something, please clarify with examples.

Reply:
It has to do with downstream parsers. Things like ElasticSearch don't know what to do with Array fields.

If you're only parsing it with other fluentd plugins, then yes, you're correct that the array is just as simple to index as a map.

However, if you're using a downstream search store, they're used to receiving data in maps rather than arrays. See https://www.elastic.co/guide/en/elasticsearch/reference/current/array.html specifically

```
Arrays of objects do not work as you would expect: you cannot query each object independently of the other objects in the array. If you need to be able to do this then you should use the [nested](https://www.elastic.co/guide/en/elasticsearch/reference/current/nested.html) data type instead of the [object](https://www.elastic.co/guide/en/elasticsearch/reference/current/object.html) data type.
```

Additionally, every other field that fluent-plugin-kubernetes_metadata_filter produces is a map, so I think that this fits the principle of least surprise when people are enabling the feature.

Reply:
Note the above document mentions "nested" fields which is a way to change the mapping in ES to handle arrays. However, this requires that the user manually apply a mapping ahead of time before enabling the feature, and it is also ES specific, finally it essentially does the same thing in the background of making it a map internally.

I would argue that just emitting it as a number indexed map is simpler all the way around unless there is some implicit cost that we're worried about with using a hash vs an array in Ruby.

Additionally, the mapping option is ES specific. Emitting it as a map in the first place is more likely to work with other logging systems out of the box similar to other fields that are already emitted by the plugin.

Reply:
If the int-map is a specific workaround for ES, would it make sense to use the array notation in the plugin here where it's not tied to a particular output and do the "ES specific" processing later in a filter before shipping the entries to the output?

Reply:
My ask was specifically geared towards what is the better API and not output specific.  The fact that the other fields are received as a map seems immaterial from an API perspective; they are truly mapped entities and the representation should, IMO, represent how they are received.  I would prefer to not design the API with Elasticsearch in mind.  There are tools (e.g. jq) that are capable of filtering a nested array.

> I would argue that just emitting it as a number indexed map is simpler all the way around unless there is some implicit cost that we're worried about with using a hash vs an array in Ruby.

This is an API question and nothing to do with the cost in Ruby.  Ruby is quite good at spinning through enumerations and I suspect whether it is `map.each do |k,v|` or `array.each_with_index do |v,i|` they are roughly the same

Reply:
My thought is that there are likely other log stores that behave similarly although I don't have much experience with other stores. I can definitely say that reducing it to an array would be much harder on our use case, but I'm unable to say that it would be that way for all users. I usually tend towards being more pragmatic with this type of thing.

Everyone else in this thread seems to think Arrays would be better, so I'm willing to make that change if that's what would make this work mergable. Data in a bad format for our use-case would still be better than no data at all.

Reply:
I have reverted the data type to be an array and updated the tests to resolve the concerns that have been mentioned here.

Please let me know if there are any other required changes to make this mergeable.